File: sdks/js_sdk/src/clients/scheduler.ts
import * as grpc from '@grpc/grpc-js';
import { BaseClient, ClientOptions } from '../internal/baseClient.js';
import { msToDuration } from '../internal/time.js';

type SchedulerServiceClient = grpc.Client & {
  SubmitTask(req: SubmitTaskRequest, meta: grpc.Metadata, opts: grpc.CallOptions, cb: (err: grpc.ServiceError | null, res: SubmitTaskResponse) => void): void;
  RequestPreemption(req: PreemptRequest, meta: grpc.Metadata, opts: grpc.CallOptions, cb: (err: grpc.ServiceError | null, res: PreemptResponse) => void): void;
  ShutdownAgent(req: ShutdownAgentRequest, meta: grpc.Metadata, opts: grpc.CallOptions, cb: (err: grpc.ServiceError | null, res: ShutdownAgentResponse) => void): void;
  PollActivityBuffer(req: PollActivityBufferRequest, meta: grpc.Metadata, opts: grpc.CallOptions, cb: (err: grpc.ServiceError | null, res: PollActivityBufferResponse) => void): void;
  PurgeActivity(req: PurgeActivityRequest, meta: grpc.Metadata, opts: grpc.CallOptions, cb: (err: grpc.ServiceError | null, res: PurgeActivityResponse) => void): void;
};

// Types aligned with protos/scheduler.proto (dynamic loader shapes)
export interface SubmitTaskRequest {
  agent_id: string;
  task_id: string;
  priority: number; // -19..20
  params: Uint8Array;
  content_type: string;
  scope: string;
}
export interface SubmitTaskResponse { accepted: boolean; reason?: string }

export interface PreemptRequest { agent_id: string; task_id: string; reason?: string }
export interface PreemptResponse { enqueued: boolean }

export interface ShutdownAgentRequest { agent_id: string; grace_period?: { seconds: number | string; nanos?: number } }
export interface ShutdownAgentResponse { ok: boolean }

export interface PollActivityBufferRequest { agent_id: string }
export interface ActivityEntry { task_id: string; repo_id: string; worktree_id: string; branch: string; description: string; timestamp: string }
export interface PollActivityBufferResponse { entries: ActivityEntry[] }

export interface PurgeActivityRequest { agent_id: string; task_ids: string[] }
export interface PurgeActivityResponse { purged: number }

export class SchedulerClient extends BaseClient {
  private client: SchedulerServiceClient;

  constructor(opts: ClientOptions) {
    super(opts);
    this.client = this.getServiceClient<SchedulerServiceClient>('sw4rm.scheduler.SchedulerService');
  }

  async submitTask(req: SubmitTaskRequest): Promise<SubmitTaskResponse> {
    validatePriority(req.priority);
    const meta = this.metadata();
    const opts: grpc.CallOptions = { deadline: this.deadlineFromNow() };
    return this.withRetryUnary(() => new Promise<SubmitTaskResponse>((resolve, reject) => {
      this.client.SubmitTask(req, meta, opts, (err, res) => {
        if (err) return reject(err);
        resolve({ accepted: !!res?.accepted, reason: res?.reason });
      });
    }), 'sw4rm.scheduler.SubmitTask', meta);
  }

  async requestPreemption(agentId: string, taskId: string, reason?: string): Promise<PreemptResponse> {
    const meta = this.metadata();
    const opts: grpc.CallOptions = { deadline: this.deadlineFromNow() };
    const req: PreemptRequest = { agent_id: agentId, task_id: taskId, reason };
    return this.withRetryUnary(() => new Promise<PreemptResponse>((resolve, reject) => {
      this.client.RequestPreemption(req, meta, opts, (err, res) => {
        if (err) return reject(err);
        resolve({ enqueued: !!res?.enqueued });
      });
    }), 'sw4rm.scheduler.RequestPreemption', meta);
  }

  async shutdownAgent(agentId: string, graceMs: number): Promise<ShutdownAgentResponse> {
    const meta = this.metadata();
    const opts: grpc.CallOptions = { deadline: this.deadlineFromNow() };
    const req: ShutdownAgentRequest = { agent_id: agentId, grace_period: msToDuration(graceMs) };
    return this.withRetryUnary(() => new Promise<ShutdownAgentResponse>((resolve, reject) => {
      this.client.ShutdownAgent(req, meta, opts, (err, res) => {
        if (err) return reject(err);
        resolve({ ok: !!res?.ok });
      });
    }), 'sw4rm.scheduler.ShutdownAgent', meta);
  }

  async pollActivityBuffer(agentId: string): Promise<ActivityEntry[]> {
    const meta = this.metadata();
    const opts: grpc.CallOptions = { deadline: this.deadlineFromNow() };
    const req: PollActivityBufferRequest = { agent_id: agentId };
    return this.withRetryUnary(() => new Promise<ActivityEntry[]>((resolve, reject) => {
      this.client.PollActivityBuffer(req, meta, opts, (err, res) => {
        if (err) return reject(err);
        resolve(res?.entries ?? []);
      });
    }), 'sw4rm.scheduler.PollActivityBuffer', meta);
  }

  async purgeActivity(agentId: string, taskIds: string[]): Promise<number> {
    const meta = this.metadata();
    const opts: grpc.CallOptions = { deadline: this.deadlineFromNow() };
    const req: PurgeActivityRequest = { agent_id: agentId, task_ids: taskIds };
    return this.withRetryUnary(() => new Promise<number>((resolve, reject) => {
      this.client.PurgeActivity(req, meta, opts, (err, res) => {
        if (err) return reject(err);
        resolve(Number(res?.purged ?? 0));
      });
    }), 'sw4rm.scheduler.PurgeActivity', meta);
  }
}

function validatePriority(p: number) {
  if (!Number.isFinite(p) || p < -19 || p > 20) {
    throw new Error('priority must be in range -19..20');
  }
}

File: examples/reference-services/rust/src/scheduler.rs
use crate::proto::sw4rm::scheduler::{
    scheduler_service_server::{SchedulerService, SchedulerServiceServer},
    SubmitTaskRequest, SubmitTaskResponse,
    PreemptRequest, PreemptResponse,
    ShutdownAgentRequest, ShutdownAgentResponse,
    PollActivityBufferRequest, PollActivityBufferResponse,
    PurgeActivityRequest, PurgeActivityResponse,
    ActivityEntry,
};
use dashmap::DashMap;
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};
use tonic::{Request, Response, Status};
use tracing::{info, debug};
use uuid::Uuid;

#[derive(Debug, Clone)]
pub struct TaskInfo {
    pub task_id: String,
    pub agent_id: String,
    pub priority: i32,
    pub params: Vec<u8>,
    pub content_type: String,
    pub scope: String,
    pub submitted_at: u64,
}

#[derive(Debug)]
pub struct SchedulerServiceImpl {
    tasks: Arc<DashMap<String, TaskInfo>>,
    activity_buffer: Arc<DashMap<String, Vec<ActivityEntry>>>,
}

impl SchedulerServiceImpl {
    pub fn new() -> Self {
        info!("Scheduler service initialized");
        Self {
            tasks: Arc::new(DashMap::new()),
            activity_buffer: Arc::new(DashMap::new()),
        }
    }
}

#[tonic::async_trait]
impl SchedulerService for SchedulerServiceImpl {
    async fn submit_task(
        &self,
        request: Request<SubmitTaskRequest>,
    ) -> Result<Response<SubmitTaskResponse>, Status> {
        let req = request.into_inner();
        let task_id = if req.task_id.is_empty() {
            Uuid::new_v4().to_string()
        } else {
            req.task_id
        };

        let current_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();

        let task_info = TaskInfo {
            task_id: task_id.clone(),
            agent_id: req.agent_id.clone(),
            priority: req.priority,
            params: req.params,
            content_type: req.content_type,
            scope: req.scope,
            submitted_at: current_time,
        };

        self.tasks.insert(task_id.clone(), task_info);
        info!("Submitted task {} for agent {}", task_id, req.agent_id);

        Ok(Response::new(SubmitTaskResponse {
            accepted: true,
            reason: format!("Task {} submitted successfully", task_id),
        }))
    }

    async fn request_preemption(
        &self,
        request: Request<PreemptRequest>,
    ) -> Result<Response<PreemptResponse>, Status> {
        let req = request.into_inner();
        info!("Preemption requested for task {} (agent {}): {}", req.task_id, req.agent_id, req.reason);
        
        // For this simple implementation, we'll just acknowledge the preemption
        Ok(Response::new(PreemptResponse { enqueued: true }))
    }

    async fn shutdown_agent(
        &self,
        request: Request<ShutdownAgentRequest>,
    ) -> Result<Response<ShutdownAgentResponse>, Status> {
        let req = request.into_inner();
        info!("Shutdown requested for agent {}", req.agent_id);
        
        // Remove all tasks for this agent
        let mut tasks_to_remove = Vec::new();
        for entry in self.tasks.iter() {
            if entry.value().agent_id == req.agent_id {
                tasks_to_remove.push(entry.key().clone());
            }
        }
        
        for task_id in tasks_to_remove {
            self.tasks.remove(&task_id);
        }
        
        // Clear activity buffer for this agent
        self.activity_buffer.remove(&req.agent_id);
        
        Ok(Response::new(ShutdownAgentResponse { ok: true }))
    }

    async fn poll_activity_buffer(
        &self,
        request: Request<PollActivityBufferRequest>,
    ) -> Result<Response<PollActivityBufferResponse>, Status> {
        let req = request.into_inner();
        debug!("Polling activity buffer for agent {}", req.agent_id);
        
        let entries = self.activity_buffer
            .get(&req.agent_id)
            .map(|e| e.clone())
            .unwrap_or_default();
        
        Ok(Response::new(PollActivityBufferResponse { entries }))
    }

    async fn purge_activity(
        &self,
        request: Request<PurgeActivityRequest>,
    ) -> Result<Response<PurgeActivityResponse>, Status> {
        let req = request.into_inner();
        debug!("Purging activity for agent {}, tasks: {:?}", req.agent_id, req.task_ids);
        
        let purged_count = if let Some(mut entries) = self.activity_buffer.get_mut(&req.agent_id) {
            let initial_len = entries.len();
            if req.task_ids.is_empty() {
                entries.clear();
                initial_len
            } else {
                entries.retain(|entry| !req.task_ids.contains(&entry.task_id));
                initial_len - entries.len()
            }
        } else {
            0
        };
        
        Ok(Response::new(PurgeActivityResponse { 
            purged: purged_count as u32 
        }))
    }
}

pub fn create_service() -> SchedulerServiceServer<SchedulerServiceImpl> {
    SchedulerServiceServer::new(SchedulerServiceImpl::new())
}